src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt

```diff
@@ -273,6 +264,30 @@ class ReferralService(
     return referralRepository.save(referral)
   }
 
+  // fun updateDraftReferralDesiredOutcomes(referral: Referral, desiredOutcomes: DesiredOutcomes)
+
+  fun updateDraftReferralComplexityLevel(referral: Referral, serviceCategoryId: UUID, complexityLevelId: UUID): Referral {
+    val errors = mutableListOf<FieldError>()
+
+    if (referral.serviceCategories.isEmpty()) {
+      errors.add(FieldError(field = "complexityLevel", error = Code.SERVICE_CATEGORY_MUST_BE_SET))
+    }
+
+    val validComplexityLevelIds = serviceCategoryRepository.findByIdOrNull(serviceCategoryId)?.complexityLevels?.map { it.id }
+      ?: throw ServerWebInputException("invalid service category id")
+
+    if (!validComplexityLevelIds.contains(complexityLevelId)) {
+      errors.add(FieldError(field = "complexityLevel", error = Code.INVALID_COMPLEXITY_LEVEL_FOR_SERVICE_CATEGORY))
+    }
+
+    if (errors.isNotEmpty()) {
+      throw ValidationError("draft referral update invalid", errors)
+    }
+
+    referral.complexityLevelIds[serviceCategoryId] = complexityLevelId
+    return referralRepository.save(referral)
+  }
```

Comment:
should also check that the service category is valid for the contract type too?

Reply:
hm. i think this validation lives in the method that allows setting of the service category. i think here we can be happy that the service categories which have been selected are valid.